cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -266,6 +266,52 @@ void EnsureLookupTablesFilled() {}
 
 #endif  // ARROW_WITH_UTF8PROC
 
+template <typename Type>
+struct AsciiReverse : StringTransform<Type, AsciiReverse<Type>> {
+  using Base = StringTransform<Type, AsciiReverse<Type>>;
+  using offset_type = typename Base::offset_type;
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
+    uint8_t utf8_char_found = 0;
+    for (offset_type i = 0; i < input_string_ncodeunits; i++) {
+      // if a utf8 char is found, report to utf8_char_found
+      utf8_char_found |= input[i] & 0x80;
+      output[input_string_ncodeunits - i - 1] = input[i];
+    }
+    //    todo: finalize if L278 check is required or not. If not required,
+    //    simply use the following
+    //    std::reverse_copy(input, input + input_string_ncodeunits, output);
+    *output_written = input_string_ncodeunits;
+    return utf8_char_found == 0;
+  }
+};
+
+/*
+ * UTF8 codeunit size can be determined by looking at the leading 4 bits of BYTE1
+ */
+const std::array<uint8_t, 16> UTF8_BYTE_SIZE_LUT{1, 1, 1, 1, 1, 1, 1, 1,
+                                                 0, 0, 0, 0, 2, 2, 3, 4};
```

Comment:
This deserves to be an inline function in `util/utf8.h`, I think (perhaps it even already exists?).

Reply:
I actually couldn't find it there. I can actually add an inline function like this
```c++
inline uint8_t Utf8CharByteWidth(const uint8 *char_ptr);
```
WDYT?

Reply:
There is `ARROW_FORCE_INLINE` in "util/macros.h"

Reply:
If \xFF is invalid, then it's probably ok either way if it skips 4 bytes or one, because the input is invalid and we don't care about the end result so long as we don't crash or invoke undefined behavior.

Reply:
Sorry @nirandaperera do you want to take Antoine's suggestion and move this to utf8.h as a `static inline` function? Also, please note in the docstring that it considers \xFF as the start of a 4-byte code unit but that this is intentional as this is an invalid byte sequence.

Reply:
Done